### PR TITLE
chore: support for openapi query param formats

### PR DESCRIPTION
**Pull request by OmarAlJarrah**
# Situation
The current implementation of URL query parameter handling in the SDK does not support various OpenAPI query parameter formats. This limitation affects the flexibility and compatibility of the SDK with different API specifications.

# Task
The objective is to enhance the SDK by adding support for multiple OpenAPI query parameter formats, including form, explode, space-delimited, and pipe-delimited formats. This will improve the SDK's ability to handle diverse API query parameter requirements.

# Action
- Introduced a new `UrlQueryParam` data class to represent URL query parameters.
- Implemented the `StringifyQueryParam` functional interface for converting `UrlQueryParam` to a string.
- Added utility functions (`stringifyForm`, `stringifyExplode`, `stringifySpaceDelimited`, `stringifyPipeDelimited`) to handle different query parameter formats.
- Created comprehensive unit tests for each query parameter format to ensure correct functionality.

# Testing
- Developed unit tests for the `UrlQueryParam` class to verify initialization and string conversion.
- Implemented tests for each query parameter format utility function to validate their output against expected results.
- Ensured edge cases such as empty keys, special characters, and large input sizes are covered in the tests.

# Results
- The SDK now supports multiple OpenAPI query parameter formats, enhancing its compatibility with various APIs.
- Unit tests confirm the correct implementation of the new features, ensuring reliability and robustness.
- The changes provide a more flexible and comprehensive solution for handling URL query parameters in the SDK.

# Notes
- Documentation updates may be required to reflect the new query parameter handling capabilities.
- Related issues and follow-up tasks include integrating these changes into the main SDK workflow and ensuring backward compatibility.
- Potential risks include the need for thorough testing in real-world scenarios to identify any unforeseen issues.

### CODE REVIEW

File: expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Data class representing a URL query parameter.
+ *
+ * @property key The key of the query parameter.
+ * @property value The list of values associated with the query parameter.
+ * @property stringify The stringifier used to convert the query parameter to a string.
+ */
+data class UrlQueryParam(
+    val key: String,
+    val value: List<String>,
+    val stringify: StringifyQueryParam
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt:15]: Nice!

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        val result = param.toString()
+
+        assertEquals("key=", result)
+    }
+
+    @Test
+    fun `converts to string with special characters`() {
+        val mockStringify = mockk<StringifyQueryParam>()
+        val param = UrlQueryParam("key", listOf("value1", "value&2"), mockStringify)
+        every { mockStringify.invoke(param) } returns "key=value1,value%262"
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt:53]: What are we testing here?

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Test
+    fun `converts to string with null key`() {
+        val mockStringify = mockk<StringifyQueryParam>()
+        val param = UrlQueryParam("", listOf("value1", "value2"), mockStringify)
+        every { mockStringify.invoke(param) } returns "=value1,value2"
+
+        val result = param.toString()
+
+        assertEquals("=value1,value2", result)
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt:68]: throw `IllegalArgumentException` instead?

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,408 @@
+package com.expediagroup.sdk.rest.util
+
+import com.expediagroup.sdk.rest.model.UrlQueryParam
+import io.mockk.mockk
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.Assertions.assertEquals // Import JUnit Jupiter Assertions
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt:7]:
```suggestion
import org.junit.jupiter.api.Assertions.assertEquals
```

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        fun `should return form-encoded string for key with special characters`() {
+            val param = UrlQueryParam("key with spaces", listOf("value1"), stringifyQueryParam)
+            val expected = "key with spaces=value1"
+            assertEquals(expected, stringifyForm(param))
+        }
+
+        @Test
+        fun `should return form-encoded string for values with special characters`() {
+            val param = UrlQueryParam("key1", listOf("value with,comma", "value with&ampersand", "value with=equals"), stringifyQueryParam)
+            val expected = "key1=value with,comma,value with&ampersand,value with=equals"
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt:47]: is this valid query params string?

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+        @Test
+        fun `invoke should add a prefix`() {
+            val customStringify = StringifyQueryParam { param ->
+                "prefix_${param.key}=${param.value.joinToString(",")}"
+            }
+            val param = UrlQueryParam("key", listOf("v1", "v2"), stringifyQueryParam)
+            val expected = "prefix_key=v1,v2"
+            assertEquals(expected, customStringify(param))
+        }
+    }
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt:407]: We don't need to assert anything other than it uses the provided logic (first test case), these test cases are not adding any value

File: expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package com.expediagroup.sdk.rest.model
```
Comment (Mohammad-Dwairi) [inline expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt:1]: please add license headers to the new files

File: expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt (hunk 1 above)
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/main/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParam.kt:15]: Thanks 🤟🏻

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt (hunk 3 above)
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/model/UrlQueryParamTest.kt]: done.

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt (hunk 4 above)
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt]: done.

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+        fun `should return form-encoded string for empty value list`() {
+            val param = UrlQueryParam("key1", emptyList(), stringifyQueryParam)
+            val expected = "key1="
+            assertEquals(expected, stringifyForm(param))
+        }
+
+        @Test
+        fun `should return form-encoded string for key with special characters`() {
+            val param = UrlQueryParam("key with spaces", listOf("value1"), stringifyQueryParam)
+            val expected = "key with spaces=value1"
```
[the remark this answers is not in the document]
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt:40]: After taking a look again nope. Test removed.

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt (hunk 5 above)
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt:47]: Test removed.

File: expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt (hunk 6 above)
Comment (OmarAlJarrah) [reply expediagroup-sdk-rest/src/test/kotlin/com/expediagroup/sdk/rest/util/UrlQueryParamUtilsTest.kt]: Done.

**Review by Mohammad-Dwairi — APPROVED** [review]

**Review by anssari1 — APPROVED** [review]
